File: rust/axum-demo/src/reply.rs

use axum::{response::{IntoResponse,Response}, http::StatusCode};
use once_cell::sync::Lazy;
use serde_json::{Value, json};
use serde::Serialize;

use crate::exception::Exception;

use std::collections::BTreeMap;

pub struct Reply(Value, i32);
pub type Result<T> = std::result::Result<T, Exception>;

static MAP: Lazy<BTreeMap<i32, &'static str>> = Lazy::new(|| {
    Reply::get_map()
});

impl Reply {

    //error kind
    pub const USER_EXCEPTION: i32 = -1000;
    pub const MODEL_EXCEPTION: i32 = -2000;
    pub const SERVICE_EXCEPTION: i32 = -3000;
    pub const FLUX_EXCEPTION: i32 = -4000;

    //error code
    pub const SUCCESS: i32 = 0;
    pub const BAD_RESULT: i32 = 1000;
    pub const BAD_TOKEN: i32 = 1100;
    pub const BAD_PARAM: i32 = 1200;
    pub const OPERATION_NOT_ALLOWED: i32 = 1300;
    pub const RESOURCE_NOT_FOUND: i32 = 1400;
    pub const OPERATION_FAILED: i32 = 1500;
    pub const OPERATION_PENDING: i32 = 1600;    

    pub fn new(data: Value, code: i32) -> Self {
        Reply(data, code)
    }

    pub fn get_map() -> BTreeMap<i32, &'static str> {
        BTreeMap::from([
            (401, "unauthorized"),
            (403, "forbidden"),
            (404, "not_found"),
            (405, "method_not_allowed"),
            (500, "internal_server_error"),
    
            (Self::SUCCESS, "success"),
            (Self::BAD_RESULT, "bad_result"),
            (Self::BAD_TOKEN, "bad_token"),
            (Self::BAD_PARAM, "bad_param"),
            (Self::OPERATION_NOT_ALLOWED, "operation_not_allowed"),
            (Self::RESOURCE_NOT_FOUND, "resource_not_found"),
            (Self::OPERATION_FAILED, "operation_failed"),
            (Self::OPERATION_PENDING, "operation_pending"),
        ])
    }    

    pub fn code_to_str(code: i32) -> &'static str {
        MAP.get(&code).unwrap_or(&"none")
    }
    
    pub fn code_to_kind(code: i32) -> i32 {
        if code >= 1000 {
            return Self::USER_EXCEPTION;
        }
        code - (code % 1000)
    }    

    pub fn success<T: Serialize>(resp: T) -> Self {
        let data = json!({
            "code": 0,
            "message": "success",
            "reason": "success",
            "data": resp,
        });
    
        Reply(data, 0)
    }

    pub fn result_success<T: Serialize>(resp: T) -> Result<Self> {
        Ok(Self::success(resp))
    }
    
    pub fn failed(message: &str, code: i32) -> Self {
        let data = json!({
            "code": code,
            "message": message,
            "reason": Reply::code_to_str(code),
            "data": None::<i32>,
        });
    
        Reply(data, code)
    }

    pub fn result_failed(message: &str, code: i32) -> Result<Self> {
        Ok(Self::failed(message, code))
    }

    pub fn result_error(message: &str, code: i32) -> Result<Self> {
        Err(Exception::from((code, message)))
    }

    pub fn status_code(code: i32) -> u16 {
        let mut c: i32 = 200;
        //Status code values in the range 100-999 (inclusive) are supported
        if code >= 100 && code < 1000 {
            c = code;
        }
        c as u16
    }

}

impl IntoResponse for Reply {
    fn into_response(self) -> Response {
        
        let json = self.0.to_string();
        let code = Self::status_code(self.1);
        (
            StatusCode::from_u16(code).unwrap(),
            [("content-type", "application/json")],
            json
        ).into_response()
    }
}

File: python/fapi-demo/core/reply.py
from datetime import datetime
from enum import IntEnum
import json
from datetime import timezone
from typing import Any
from fastapi import Response

from fastapi.responses import RedirectResponse
from pydantic import BaseModel

from core.time_util import fallback_to_local_timezone
from core.request_context import getRequestContext

class Reply(IntEnum):
    SUCCESS = 0
    UNAUTHORIZED = 401
    FORBIDDEN = 403
    NOT_FOUND = 404
    METHOD_NOT_ALLOWED = 405
    INTERNAL_SERVER_ERROR = 500

    BAD_RESULT = 1000
    BAD_TOKEN = 1100
    BAD_PARAM = 1200
    OPERATION_NOT_ALLOWED = 1300
    RESOURCE_NOT_FOUND = 1400
    OPERATION_FAILED = 1500
    OPERATION_PENDING = 1600


    def __str__(self):
        return self.name.lower()
    
    def __int__(self):
        return self.value


    @classmethod
    def success(cls, resp: Any):
        return cls.json_response(0, 'success', 'success', resp)

    @staticmethod
    def redirect(url: str, status_code: int = 307):
        return RedirectResponse(url=url, status_code=status_code)

    @staticmethod
    def code_to_str(code: int) -> str:
        return reply_map_reversed[code]


    @staticmethod
    def status_code(code: int) -> int:
        c: int = 200
        #Status code values in the range 100-999 (inclusive) are supported
        if code >= 100 and code < 1000 :
            c = code

        return c
    
    @classmethod
    def json_response(cls, code:int, message:str, reason:str, data:Any, extra: dict[str, Any] = {}):

        status_code = cls.status_code(code)
        request_id = None
        try:
            request_id = getRequestContext()['request_id']
        except Exception:
            pass
        content = {
            'code': code,
            'message': message,
            'reason': cls.code_to_str(code),
            'request-id': request_id,
            'data': data,
        }
        if extra:
            content.update(extra)

        content = json.dumps(content, cls=MyJsonEncoder)
        return Response(
            status_code=status_code,
            content=content,
            media_type='application/json',
        )


reply_map = {member.name.lower(): member.value for member in Reply}
reply_map_reversed = {v: k for k, v in reply_map.items()}


class MyJsonEncoder(json.JSONEncoder):

    def default(self, o):
        if isinstance(o, datetime):
            return fallback_to_local_timezone(o).isoformat()
        elif isinstance(o, BaseModel):
            if hasattr(o, 'api_dump'):
                return o.api_dump() # type: ignore
            return o.model_dump()

        return super().default(o)
    